synthesizer/src/vm/finalize.rs

```diff
@@ -28,31 +30,75 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
     ///     `Ratify::BlockReward(block_reward)` and `Ratify::PuzzleReward(puzzle_reward)`
     ///     to the front of the `ratifications` list.
     #[inline]
-    pub fn speculate<'a>(
+    pub fn speculate<'a, R: Rng + CryptoRng>(
         &self,
         state: FinalizeGlobalState,
         coinbase_reward: Option<u64>,
         candidate_ratifications: Vec<Ratify<N>>,
         candidate_solutions: &Solutions<N>,
         candidate_transactions: impl ExactSizeIterator<Item = &'a Transaction<N>>,
+        rng: &mut R,
     ) -> Result<(Ratifications<N>, Transactions<N>, Vec<N::TransactionID>, Vec<FinalizeOperation<N>>)> {
         let timer = timer!("VM::speculate");
 
+        // TODO (raychu86): Clean up this logic and remove the extra collect.
+        // If the transactions are not part of the genesis block, ensure each transaction is well-formed and unique. Abort any transactions that are not.
+        let block_store = self.block_store();
+        let candidate_transactions = candidate_transactions.collect::<Vec<_>>();
+        let (verified_transactions, verification_aborted_transactions) = match block_store
+            .find_block_height_from_state_root(block_store.current_state_root())?
+        {
+            // If the current state root does not exist in the block store, then the genesis block has not been introduced yet.
+            None => (candidate_transactions, vec![]),
+            // Verify transactions for all non-genesis cases.
+            _ => {
+                let rngs = (0..candidate_transactions.len()).map(|_| StdRng::from_seed(rng.gen())).collect::<Vec<_>>();
+                // Verify the transactions and collect the error message if there is one.
+                let checked_transactions: Vec<_> = cfg_into_iter!(candidate_transactions)
+                    .zip(rngs)
+                    .map(|(transaction, mut rng)| {
+                        let error_message = match self.check_transaction(transaction, None, &mut rng) {
+                            Ok(_) => None,
+                            Err(e) => Some(e.to_string()),
+                        };
+                        (transaction, error_message)
+                    })
+                    .collect();
+
+                // Separate the verified and aborted transactions.
+                checked_transactions.into_iter().fold(
+                    (Vec::new(), Vec::new()),
+                    |(mut verified, mut aborted), (transaction, error_message)| {
+                        match error_message {
+                            None => verified.push(transaction),
+                            Some(e) => aborted.push((transaction, e)),
+                        };
+                        (verified, aborted)
+                    },
+                )
```

Comment:
seems like https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.partition would be a good fit here

Reply:
Good point, much cleaner.